File: asset/js/charts/laicite/circulation.js
/**
 * IWAC Visualizations — Laïcité block: circulation (issue #19 D).
 *
 * Near-duplicate laïcité articles printed by different outlets.
 *
 * The question none of the other views can answer: does this coverage
 * *circulate*? A communiqué printed verbatim by eleven papers, a wire
 * dispatch picked up across a border, and eleven newsrooms independently
 * covering the same controversy are indistinguishable on a per-year item
 * count — and they are not the same finding. It bears directly on the
 * argument the dossier backs: a claim about the volume of debate is
 * weaker if the volume is one press release printed eleven times.
 *
 * Three parts, in the order the question is actually asked:
 *   1. a headline share — how much of the dossier is duplicated at all;
 *   2. duplication per decade, so the timeline's ridges can be read
 *      against it;
 *   3. the pairs themselves, because "trust me, they're duplicates" is
 *      not an argument a reader can check.
 *
 * The detection floor is the Press Reprints block's own threshold, so a
 * "reprint" means the same thing in both places.
 */
(function () {
    'use strict';

    var ns = window.IWACVis;
    if (!ns || !ns.panels) {
        console.warn('IWACVis.laicite circulation: missing panels — check load order');
        return;
    }
    var P = ns.panels;
    var L = ns.laicite = ns.laicite || {};

    function pct(n, d) { return d ? (n / d) * 100 : 0; }

    /**
     * Duplication per decade, as a share of that decade's scanned
     * articles rather than a count.
     *
     * A share, because the corpus grows steadily: raw counts would peak
     * wherever the dossier is largest and report corpus growth under
     * another name — the same reasoning that made the arenas view use
     * shares. The share is what can be laid beside the timeline's
     * Ivorian 1993–2001 ridge and answer whether it is many voices or
     * one text moving between titles.
     */
    function decadeOption(bundle) {
        var byDecade = bundle.by_decade || {};
        var decades = Object.keys(byDecade).sort();
        if (!decades.length) return P.emptyChartOption();

        var palette = (ns.getPalette && ns.getPalette()) || [];
        var R = ns.responsive;
        var option = {
            grid: (ns.chartOptions && ns.chartOptions._grid)
                ? ns.chartOptions._grid({ left: 56, top: 24, bottom: 44 })
                : { left: 56, right: 24, top: 24, bottom: 44, containLabel: true },
            tooltip: {
                trigger: 'axis',
                confine: true,
                axisPointer: { type: 'shadow' },
                formatter: function (params) {
                    if (!params || !params.length) return '';
                    var p = params[0];
                    return '<strong>' + P.escapeHtml(p.axisValue) + '</strong><br>'
                        + P.t('laicite.circulation_decade_tooltip', {
                            count: P.formatNumber(byDecade[p.axisValue] || 0)
                        });
                }
            },
            xAxis: { type: 'category', data: decades },
            yAxis: {
                type: 'value',
                name: P.t('laicite.circulation_axis'),
                nameLocation: 'end',
                nameGap: 12
            },
            series: [{
                type: 'bar',
                itemStyle: { color: palette[0] },
                data: decades.map(function (d) { return byDecade[d] || 0; })
            }]
        };
        return R && R.withMedia ? R.withMedia(option, {}) : option;
    }

    /** The outlet pairs that share copy most often. */
    function buildLinkList(bundle) {
        var links = (bundle.links || []).slice(0, 15);
        var wrap = P.el('div', 'iwac-vis-laicite-circ-links');
        if (!links.length) return wrap;
        wrap.appendChild(P.el('h5', null, P.t('laicite.circulation_links_title')));
        var list = P.el('ul', 'iwac-vis-laicite-circ-link-list');
        links.forEach(function (row) {
            var li = P.el('li', 'iwac-vis-laicite-circ-link');
            li.appendChild(P.el('span', 'iwac-vis-laicite-circ-link-pair',
                row[0] + ' ⇄ ' + row[1]));
            li.appendChild(P.el('span', 'iwac-vis-laicite-circ-link-n',
                P.t('laicite.circulation_pairs_n', {
                    count: P.formatNumber(row[2])
                })));
            list.appendChild(li);
        });
        wrap.appendChild(list);
        return wrap;
    }

    /**
     * The individual pairs. Shown, not just counted: a near-duplicate
     * claim is checkable only if the reader can open both sides and look.
     */
    function buildPairList(bundle, siteBase) {
        var pairs = bundle.pairs || [];
        var wrap = P.el('div', 'iwac-vis-laicite-circ-pairs');
        if (!pairs.length) return wrap;
        wrap.appendChild(P.el('h5', null, P.t('laicite.circulation_pairs_title')));
        wrap.appendChild(P.el('p', 'iwac-vis-panel-desc',
            P.t('laicite.circulation_pairs_desc', {
                listed: P.formatNumber(bundle.listed || pairs.length),
                total: P.formatNumber(bundle.total_pairs || pairs.length)
            })));

        var list = P.el('ul', 'iwac-vis-laicite-circ-pair-list');
        pairs.forEach(function (pair) {
            var li = P.el('li', 'iwac-vis-laicite-circ-pair');
            li.appendChild(P.el('span', 'iwac-vis-laicite-circ-sim',
                P.t('laicite.circulation_similarity', {
                    value: (pair.similarity * 100).toFixed(1)
                })));
            ['a', 'b'].forEach(function (side) {
                var item = pair[side] || {};
                var row = P.el('div', 'iwac-vis-laicite-circ-side');
                var title = item.title || '';
                if (siteBase && item.o_id) {
                    var link = P.el('a', 'iwac-vis-laicite-circ-title', title);
                    link.href = siteBase + '/item/' + item.o_id;
                    row.appendChild(link);
                } else {
                    row.appendChild(P.el('span', 'iwac-vis-laicite-circ-title', title));
                }
                var meta = [item.newspaper, item.year ? String(item.year) : '']
                    .filter(Boolean).join(' · ');
                if (meta) row.appendChild(P.el('span', 'iwac-vis-laicite-circ-meta', meta));
                li.appendChild(row);
            });
            list.appendChild(li);
        });
        wrap.appendChild(list);
        return wrap;
    }

    /**
     * @param {Object} cfg {bundle, siteBase}
     * @returns {{root: HTMLElement, mount: function():void}}
     */
    L.buildCirculation = function (cfg) {
        var bundle = cfg.bundle || {};
        var root = P.el('div', 'iwac-vis-laicite-circulation');
        var mounts = [];

        var panel = P.el('div', 'iwac-vis-panel');
        panel.appendChild(P.el('h4', null, P.t('laicite.circulation_title')));
        panel.appendChild(P.el('p', 'iwac-vis-panel-desc',
            P.t('laicite.circulation_desc')));

        var scanned = bundle.scanned || 0;
        var reprinted = bundle.reprinted_items || 0;

        if (!scanned) {
            panel.appendChild(P.buildNoDataState());
            root.appendChild(panel);
            return { root: root, mount: function () {} };
        }

        // Headline first: one number that answers the question, with both
        // sides of the fraction on screen so the share is checkable.
        panel.appendChild(P.buildSummaryCards([
            { value: reprinted, labelKey: 'laicite.circulation_kpi_items' },
            { value: scanned, labelKey: 'laicite.circulation_kpi_scanned' },
            { value: pct(reprinted, scanned).toFixed(1) + '%', text: true,
              labelKey: 'laicite.circulation_kpi_share' },
            { value: bundle.total_pairs || 0, labelKey: 'laicite.circulation_kpi_pairs' }
        ]));

        // The floor and its blind spot, stated where the numbers are —
        // this counts only pairs where BOTH copies made the dossier, so
        // it is a floor and never a census.
        panel.appendChild(P.el('p', 'iwac-vis-panel-desc iwac-vis-laicite-circ-note',
            P.t('laicite.circulation_note', {
                threshold: ((bundle.threshold || 0) * 100).toFixed(0)
            })));

        if (!reprinted) {
            // A real and reportable answer, not an empty state: this
            // coverage is not syndicated copy.
            panel.appendChild(P.el('p', 'iwac-vis-laicite-circ-none',
                P.t('laicite.circulation_none')));
            root.appendChild(panel);
            return { root: root, mount: function () {} };
        }

        root.appendChild(panel);

        var decadePanel = P.buildPanel('iwac-vis-panel iwac-vis-laicite-circ-decade',
            P.t('laicite.circulation_decade_title'),
            P.t('laicite.circulation_decade_desc'));
        root.appendChild(decadePanel.panel);
        mounts.push(function () {
            ns.registerChart(decadePanel.chart, function (el, instance) {
                instance.setOption(decadeOption(bundle), { notMerge: true });
            });
        });

        var listPanel = P.el('div', 'iwac-vis-panel');
        listPanel.appendChild(buildLinkList(bundle));
        listPanel.appendChild(buildPairList(bundle, cfg.siteBase));
        root.appendChild(listPanel);

        return {
            root: root,
            mount: function () { mounts.forEach(function (fn) { fn(); }); }
        };
    };
})();

File: scripts/laicite/circulation.py
"""Near-duplicate cross-outlet pairs — ``laicite-circulation.json``.
"""
from __future__ import annotations

from collections import Counter, defaultdict
from typing import Any, Dict, List, Optional, Set

from iwac_embeddings import pairs_above_threshold
from iwac_utils import generate_timestamp

from laicite.scan import ItemScan


# -----------------------------------------------------------------------
# Circulation (issue #19 D)
# -----------------------------------------------------------------------
# 0.97 is generate_reprints.py's publication threshold, kept identical so
# "reprint" means the same thing in both blocks. That script re-derives
# the cut-off empirically on every data refresh (it logs a similarity
# histogram from 0.90 up); if it ever moves, move this with it.
CIRCULATION_THRESHOLD = 0.97
# Cap on the pair list the bundle ships as browsable rows. Every
# aggregate — the outlet network, the per-decade counts, the reprinted-item
# total — is computed over ALL detected pairs before this cut, so the cap
# shortens a list without ever shrinking a number.
CIRCULATION_MAX_LISTED = 300


class CirculationMixin:
    """``LaiciteGenerator``'s circulation half. Mixed in by ``laicite.generator``."""


    def build_circulation(self) -> Dict[str, Any]:
        """Near-duplicate laïcité articles across different outlets.

        The question none of the other views can answer: does this
        coverage **circulate**? A communiqué reprinted verbatim by eleven
        papers, a PANA dispatch picked up across a border, and eleven
        newsrooms independently covering the same controversy are
        indistinguishable on a per-year item count — and they are not the
        same finding. It matters directly for the argument the dossier
        backs: a claim about the volume of debate is weaker if the volume
        is one press release printed eleven times.

        Method is ``generate_reprints.py``'s, scoped to the dossier:
        cosine similarity over L2-normalised ``embedding_OCR``, publish
        pairs above a high threshold whose two articles carry different
        newspaper names. Scoped rather than joined because that block's
        published bundle is capped at its top-N pairs by similarity and
        would silently under-report a small slice; and because ~1,300
        members is a cheap all-pairs scan where the full corpus is not.

        **A within-dossier scan can only see reprints where BOTH copies
        are members.** In practice a verbatim reprint of a laïcité
        article matches the same lexicon and joins the dossier too, so
        the pairs are near-complete — but a copy whose OCR is too poor to
        match would be missed, and the panel says the count is a floor
        rather than a census.
        """
        X, scans = self._member_embeddings()
        if len(scans) < 2:
            return {
                "generated_at": generate_timestamp(),
                "threshold": CIRCULATION_THRESHOLD,
                "scanned": len(scans),
                "listed": 0,
                "total_pairs": 0,
                "pairs": [],
                "links": [],
                "newspapers": [],
                "reprinted_items": 0,
                "median_year_gap": None,
                "by_decade": {},
            }

        pairs: List[Dict[str, Any]] = []
        for i, j, sim in pairs_above_threshold(X, CIRCULATION_THRESHOLD):
            a, b = scans[i], scans[j]
            # Same outlet is not circulation — it is a correction, a
            # second edition, or the same piece indexed twice.
            if not a.newspaper or not b.newspaper or a.newspaper == b.newspaper:
                continue
            pairs.append({
                "similarity": round(sim, 4),
                "a": self._circulation_side(a),
                "b": self._circulation_side(b),
                "year_gap": self._year_gap(a, b),
            })

        pairs.sort(key=lambda p: -p["similarity"])

        # Aggregates run over EVERY detected pair, before the display cap
        # below: a "12% of the dossier is reprinted copy" figure computed
        # from a truncated list would be wrong in the one direction a
        # reader cannot detect.
        link_counts: Dict[frozenset, int] = defaultdict(int)
        paper_counts: Counter = Counter()
        reprinted: Set[str] = set()
        by_decade: Counter = Counter()
        for p in pairs:
            pa, pb = p["a"]["newspaper"], p["b"]["newspaper"]
            link_counts[frozenset((pa, pb))] += 1
            paper_counts[pa] += 1
            paper_counts[pb] += 1
            reprinted.add(p["a"]["o_id"])
            reprinted.add(p["b"]["o_id"])
            for side in ("a", "b"):
                decade = self._decade(p[side]["year"])
                if decade:
                    by_decade[decade] += 1

        links = []
        for pair, count in sorted(link_counts.items(),
                                  key=lambda kv: (-kv[1], sorted(kv[0]))):
            left, right = sorted(pair)
            links.append([left, right, count])

        gaps = sorted(p["year_gap"] for p in pairs if p["year_gap"] is not None)
        median_gap = gaps[len(gaps) // 2] if gaps else None

        self.logger.info(
            "  circulation: %d cross-outlet pairs ≥ %.2f over %d embedded "
            "articles, touching %d items",
            len(pairs), CIRCULATION_THRESHOLD, len(scans), len(reprinted))

        return {
            "generated_at": generate_timestamp(),
            "threshold": CIRCULATION_THRESHOLD,
            # The denominator every share on the panel is taken against.
            "scanned": len(scans),
            "listed": min(len(pairs), CIRCULATION_MAX_LISTED),
            "total_pairs": len(pairs),
            "reprinted_items": len(reprinted),
            "median_year_gap": median_gap,
            "by_decade": dict(sorted(by_decade.items())),
            "newspapers": [
                {"name": name, "pairs": int(count)}
                for name, count in sorted(paper_counts.items(),
                                          key=lambda kv: (-kv[1], kv[0]))
            ],
            "links": links,
            "pairs": pairs[:CIRCULATION_MAX_LISTED],
        }

    @staticmethod
    def _circulation_side(scan: ItemScan) -> Dict[str, Any]:
        return {
            "o_id": scan.o_id,
            "title": scan.title,
            "newspaper": scan.newspaper,
            "country": scan.countries[0] if scan.countries else "",
            "year": scan.year,
        }

    @staticmethod
    def _year_gap(a: ItemScan, b: ItemScan) -> Optional[int]:
        """Years between two items, or None when either is undated.

        Years rather than days: the scan works off ``ItemScan``, which
        keeps the parsed year and not the raw date, and a great many of
        these items are ``YYYY``-only anyway. A day-level gap would be
        precision the dossier's dates do not support.
        """
        if a.year is None or b.year is None:
            return None
        return abs(a.year - b.year)

    def _dominant_annotation_frame(self, scan: ItemScan) -> str:
        """The frame that best characterises this item, or "".

        Membership frames are excluded: an item is in the dossier
        *because* it says laïcité, so colouring by `laicite` would paint
        almost every point one colour and encode the selection criterion
        instead of anything about the item. The arenas view excludes them
        from its small multiples for the same reason.
        """
        membership = set(self.lex.membership_frames)
        counts = {f: n for f, n in scan.frame_counts.items()
                  if f not in membership and n > 0}
        if not counts:
            return ""
        return max(sorted(counts), key=lambda f: counts[f])
